### ISSUE TITLE
How to adjust learning rate as model trains

### ISSUE THREAD

**Issue by delta-6400** [ibab/tensorflow-wavenet#267]
Hi all --

Sorry for what might be an obvious question for some -

How do you go about adjusting the learning rate with the number of elapsed steps?  I assume it's by modifying some stuff in the train.py file - I'm just not exactly sure what as I have zero experience with tensorflow.  

It looks to me like the actual execution of each training generation happens within the for loop that begins on line 295, with the `sess.run( ... )` command.  

I'm assuming I would do something that would check against which step we are on in the for loop, and then  make a new `optim` (as seen on lines 256-60)?  Then pass that in to sess.run...

Is this correct?

Let me know.  Thanks.

s

**Comment by vijay-ravichandran** [comment]
Yes, the actual execution begins with sess.run(). Everything leading up to it can be understood as declaration of operations. And `sess.run(...)` executes these operations.
So, you should be able to change your learning rate on the fly based on any logic such as `step % 10000`

**Comment by delta-6400** [comment]
@vijay-ravichandran 

Thanks for the response vijay!  I think I got it working :)

**Comment by lemonzi** [comment]
@delta-6400 Having a placeholder for the learning rate is easy and works. However, there are built-in functions that implement different strategies:`tf.train.exponential_decay()`, `tf.train.inverse_time_decay()`, `tf.train.piecewise_constant()`... I think they are TensorFlow-native and use Variables to keep state.

**Comment by cjonesy67** [comment]
@delta-6400 
I am also interested in modifying the learning rate as you describe, but am running into various problems (relatively new to tensorflow) . Can you provide a little more detail on how you made a new optim with the modified learning rate and passed that to sess.run in train.py? I seem to be missing something important. 

Thanks.

**Comment by lemonzi** [comment]
@cjonesy67 The `sess.run` method looks like:

```python
learning_rate_placeholder = tf.placeholder(tf.float32, [], name='learning_rate')
optimizer = tf.train.RMSPropOptimizer(learning_rate=learning_rate_placeholder)
train_op = optimizer.minimize(loss)
for step in range(max_steps):
    learning_rate = 1e-3 if step < 1000 else 1e-4
    sess.run(train_op, feed_dict={learning_rate_placeholder: learning_rate})
```

You would need to edit line 256 as well as the main loop at the bottom of `train.py`.

**Comment by tomsmith82** [comment]
I am not very experienced with Python or Tensorflow, can someone tell me in which lines I need to make what changes?

**Comment by lemonzi** [comment]
@tomsmith82 That would be https://github.com/ibab/tensorflow-wavenet/blob/master/train.py#L256 (my first two lines in the snippet) and https://github.com/ibab/tensorflow-wavenet/blob/master/train.py#L302 (my last two lines in the snippet), the rest are just for context.

**Comment by lemonzi** [comment]
The important bits for the second and last lines are `learning_rate=learning_rate_placeholder` and `feed_dict={learning_rate_placeholder: learning_rate}`.
